x-pack/plugins/infra/server/lib/log_analysis/log_entry_anomalies.ts

```diff
@@ -0,0 +1,391 @@
+
+  const jobId = getJobId(context.infra.spaceId, sourceId, 'log-entry-rate');
+
+  const {
+    mlJob,
+    timing: { spans: fetchMlJobSpans },
+  } = await fetchMlJob(context.infra.mlAnomalyDetectors, jobId);
+
+  const customSettings = decodeOrThrow(jobCustomSettingsRT)(mlJob.custom_settings);
+  const indices = customSettings?.logs_source_config?.indexPattern;
+  const timestampField = customSettings?.logs_source_config?.timestampField;
+  const tiebreakerField = sourceConfiguration.configuration.fields.tiebreaker;
+
+  if (indices == null || timestampField == null) {
+    throw new InsufficientLogAnalysisMlJobConfigurationError(
+      `Failed to find index configuration for ml job ${jobId}`
+    );
+  }
+
+  const {
+    examples,
+    timing: { spans: fetchLogEntryExamplesSpans },
+  } = await fetchLogEntryExamples(
+    context,
+    sourceId,
+    indices,
+    timestampField,
+    tiebreakerField,
+    startTime,
+    endTime,
+    dataset,
+    exampleCount,
+    callWithRequest,
+    categoryId
+  );
+
+  const logEntryExamplesSpan = finalizeLogEntryExamplesSpan();
+
+  return {
+    data: examples,
+    timing: {
+      spans: [logEntryExamplesSpan, ...fetchMlJobSpans, ...fetchLogEntryExamplesSpans],
+    },
+  };
+}
+
+export async function fetchLogEntryExamples(
+  context: RequestHandlerContext & { infra: Required<InfraRequestHandlerContext> },
+  sourceId: string,
+  indices: string,
+  timestampField: string,
+  tiebreakerField: string,
+  startTime: number,
+  endTime: number,
+  dataset: string,
+  exampleCount: number,
+  callWithRequest: KibanaFramework['callWithRequest'],
+  categoryId?: string
+) {
+  const finalizeEsSearchSpan = startTracingSpan('Fetch log rate examples from ES');
+
+  let categoryQuery: string | undefined;
+
+  // Examples should be further scoped to a specific ML category
+  if (categoryId) {
+    const parsedCategoryId = parseInt(categoryId, 10);
+
+    const logEntryCategoriesCountJobId = getJobId(
+      context.infra.spaceId,
+      sourceId,
+      logEntryCategoriesJobTypes[0]
+    );
+
+    const { logEntryCategoriesById } = await fetchLogEntryCategories(
+      context,
+      logEntryCategoriesCountJobId,
+      [parsedCategoryId]
+    );
+
+    const category = logEntryCategoriesById[parsedCategoryId];
+
+    if (category == null) {
+      throw new UnknownCategoryError(parsedCategoryId);
+    }
+
+    categoryQuery = category._source.terms;
+  }
+
+  const {
+    hits: { hits },
+  } = decodeOrThrow(logEntryExamplesResponseRT)(
+    await callWithRequest(
+      context,
+      'search',
+      createLogEntryExamplesQuery(
+        indices,
+        timestampField,
+        tiebreakerField,
+        startTime,
+        endTime,
+        dataset,
+        exampleCount,
+        categoryQuery
+      )
+    )
+  );
+
+  const esSearchSpan = finalizeEsSearchSpan();
+
+  return {
+    examples: hits.map((hit) => ({
+      id: hit._id,
+      dataset,
```

Comment:
Do we want to use the actual dataset just in case the semantics of the function argument and the result diverge?

```suggestion
      dataset: hit._source.event?.dataset ?? '',
```

Reply:
Yes, good point.